src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,368 @@
+
+  mPredicates = parameterAsEnums( parameters, QStringLiteral( "PREDICATE" ), context );
+
+  QString prefix = parameterAsString( parameters, QStringLiteral( "PREFIX" ), context );
+
+  if ( mFieldNames.empty() )
+  {
+    outFields2 = mJoinSource->fields();
+    fields2Indices.reserve( outFields2.count() );
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      fields2Indices << i;
+    }
+  }
+  else
+  {
+    fields2Indices.reserve( mFieldNames.count() );
+    for ( const QString &field : mFieldNames )
+    {
+      int index = mJoinSource->fields().lookupField( field );
+      if ( index >= 0 )
+      {
+        fields2Indices << index;
+        outFields2.append( mJoinSource->fields().at( index ) );
+      }
+    }
+  }
+
+  if ( !prefix.isEmpty() )
+  {
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      outFields2[ i ].setName( prefix + outFields2[ i ].name() );
+    }
+  }
+
+  if ( feedback->isCanceled() )
+    return false;
+
+  for ( int i = 0; i < fields2Indices.count(); ++i )
+    emptyAttrs << QVariant();
+
+  fields2Fetch = fields2Indices;
+  fields2Fetch << fields2Indices;
+
+  static const QgsFields mOutFields = QgsProcessingUtils::combineFields( mBaseSource->fields(), outFields2 );
+
+  joinedFeatures.reset( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, joinedFeaturesId, mOutFields,
+                                         mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+
+  if ( parameters.value( QStringLiteral( "OUTPUT" ) ).isValid() && !joinedFeatures )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  mDiscardNonMatching = parameterAsBoolean( parameters, QStringLiteral( "DISCARD_NONMATCHING" ), context );
+
+  if ( !mDiscardNonMatching )
+    {
+    unjoinedFeatures.reset( parameterAsSink( parameters, QStringLiteral( "NON_MATCHING" ), context, unjoinedFeaturesId, mBaseSource->fields(),
+                            mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+    if ( parameters.value( QStringLiteral( "NON_MATCHING" ) ).isValid() && !unjoinedFeatures )
+      throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "NON_MATCHING" ) ) );
+    }
+
+  return true;
+}
+
+
+QVariantMap QgsJoinByLocationAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  Q_UNUSED( parameters );
+  int joinedCount = 0;
+  QgsFeatureIterator joinIter = mJoinSource->getFeatures( QgsFeatureRequest().setDestinationCrs( mBaseSource->sourceCrs(), context.transformContext() ).setSubsetOfAttributes( fields2Fetch ) );
+  static QHash< QgsFeatureId, QgsAttributes > input2AttributeCache;
+  QgsFeature f;
+  double step = mJoinSource->featureCount() > 0 ? 50.0 / mJoinSource->featureCount() : 1;
+  int i = 0;
```

Comment:
From here the c++ differs from the original python version -- can you adapt the python version as is for now?

Reply:
This loop was removed, remaining should be the same in function.